sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/PlanQueryStage.scala

```diff
@@ -0,0 +1,57 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.adaptive
+
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.execution.SparkPlan
+import org.apache.spark.sql.execution.exchange.{BroadcastExchangeExec, Exchange, ReusedExchangeExec, ShuffleExchangeExec}
+
+/**
+ * Divide the spark plan into multiple QueryStages. For each Exchange in the plan, it wraps it with
+ * a [[QueryStage]]. At the end it adds an [[AdaptiveSparkPlan]] at the top, which will drive the
+ * execution of query stages.
+ */
+case class PlanQueryStage(session: SparkSession) extends Rule[SparkPlan] {
+
+  def apply(plan: SparkPlan): SparkPlan = {
+    var id = 0
+    val exchangeToQueryStage = new java.util.IdentityHashMap[Exchange, QueryStage]
+    val planWithStages = plan.transformUp {
+      case e: ShuffleExchangeExec =>
+        val queryStage = ShuffleQueryStage(id, e)
```

Comment:
Just a question; I'm not 100% sure about the design though... during runtime, a query stage could change depends on the map output statistics of the child stages, e.g., from `ShuffleQueryStage` to `BroadcastExchangeExec`?

Reply:
Yes, you can preview the relative code here, as our plan, the join optimization rule will in a separated PR.
https://github.com/Intel-bigdata/spark-adaptive/blob/ae-2.3-08/sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeJoin.scala#L114

Reply:
So, this might be a comment about the design though, we can't do planning from a logical plan to a physical plan in a query stage on-runtime instead of the current design (replacing from a already-planned physical plan to another physical plan on-runtime)?